### PR TITLE
Lut selection Insight (2)

### PR DESCRIPTION
**Pull request by dominikl**
# What this PR does

Adds Lookup Tables support to Insight.

# Testing this PR

Select an Image. Go to right hand side Preview panel. Make sure you can select the lookup tables in the Channel color picker.

### CODE REVIEW

File: components/insight/SRC/org/openmicroscopy/shoola/util/ui/colourpicker/LookupTableItem.java [hunk 1]
```diff
@@ -99,6 +125,46 @@ public boolean matchesFilename(String filename) {
...
+            return false;
+        if (label == null) {
+            if (other.label != null)
+                return false;
+        } else if (!label.equals(other.label))
+            return false;
+        return true;
     }
+    
+
```
Comment (mtbc) [inline components/insight/SRC/org/openmicroscopy/shoola/util/ui/colourpicker/LookupTableItem.java]: not sure we need these blanks

File: components/insight/SRC/org/openmicroscopy/shoola/util/ui/colourpicker/LookupTableItem.java [hunk 2]
```diff
@@ -99,6 +125,46 @@ public boolean matchesFilename(String filename) {
 
     @Override
     public int compareTo(LookupTableItem o) {
-        return this.readableName.compareTo(o.readableName);
+        return this.label.compareTo(o.label);
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
```
Comment (mtbc) [inline components/insight/SRC/org/openmicroscopy/shoola/util/ui/colourpicker/LookupTableItem.java]: FYI the top of https://github.com/google/guava/wiki/CommonObjectUtilitiesExplained has some nice stuff offering convenience.

File: components/insight/SRC/org/openmicroscopy/shoola/util/ui/colourpicker/LookupTableListRenderer.java [hunk 3]
```diff
@@ -22,32 +22,81 @@
...
+
+        if (isSelected) {
+            setForeground(list.getSelectionForeground());
+            setBackground(list.getSelectionBackground());
+        } else {
+            setForeground(list.getForeground());
+            setBackground(list.getBackground());
+        }
+        
+        if (cellHasFocus) {
```
Comment (mtbc) [inline components/insight/SRC/org/openmicroscopy/shoola/util/ui/colourpicker/LookupTableListRenderer.java]: ternary operator may be nice here, unless we eschew it; turn six lines to one

**Comment by dominikl** [comment]
Thanks @mtbc , added your suggestions.

**Comment by will-moore** [comment]
I wrote a quick python script to generate Preview of all our know LUTs.
luts_10

This is what I'm using to use to show LUT preview in the UI
screen shot 2016-07-29 at 16 16 17

Script is at https://gist.github.com/will-moore/49e0d1ce175e624184c362fed3441992 if you want to adapt to generate suitable icon or multiple icons for Insight?

**Comment by dominikl** [comment]
Using icons for the LUTs (and as background for the channel slider) definitely looks better. I think using @will-moore 's pre-generated PNG for now is a good solution, until this maybe made more flexible and moved to the server side.

**Comment by dominikl** [comment]
Test with breaking server (see #4753)

**Comment by will-moore** [comment]
UI looks nice with the icons. Works well too. Channel buttons and slider update as expected. Button text is slightly tricky to read, but I think this is OK (can't think of a nice solution and likely only a single channel will have LUT at any time).

If I pick a simple color, E.g. Red, then the next time I open the color-picker dialog, I get the full color-picker wheel, instead of the list of colors & LUTs that I used to pick the Red (and user may not know how to return to this list). Either need to check if current color is in the list of preset colors, OR always show the list first (at least the user will know how to find the full picker wheel).

Theres's a bug with the Alpha slider (which I've never used before). This always shows the last selected color, and if you drag the slider then that color is also shown at the top of the dialog and is picked if you "Accept".

I'm not sure if Alpha is used much, but it can't be used with LUTs, right @jburel? So probably just need to disable it when LUT is selected. We don't support alpha at-all in web.

screen shot 2016-08-02 at 10 16 48

**Comment by jburel** [comment]
the alpha component is not taken into account when using a lut.

**Comment by dominikl** [comment]
Yes, you should only see the color wheel as default, when the channel color is not in the list, that's already implemented, I'll have to check that again.

**Comment by dominikl** [comment]
I'll disable the alpha slider then, when a lut is selected.

**Comment by will-moore** [comment]
On second thoughts, setting the channel button to a generic color instead of LUT background might work quite well, since you always have the slider itself to show the LUT:

screen shot 2016-08-02 at 11 28 07

**Comment by gusferguson** [comment]
@dominikl 

Tested using OMERO.insight-5.3.0-m2-1002-6b855b6-ice36-b386-mac user-3 read-only-1

Selection of LUT as channel colour is not changing the colour of the channel at all in either Preview or Full Viewer or thumbnails - presuming the backend has not been implemented yet.

With all selections - colour, LUT, color picker, all save, save to all, copy, paste and undo are working as expected in Preview, Full Viewer and data tree right-click menu

Background of buttons to left of slider are all appearing with lighter end of LUT selected - I agree with Will that the grey BG is probably better as the LUT is represented in the slider colour.
In the full-viewer the left hand buttons' background representing the LUT is a good compromise as there are no sliders.

Issues:
On selected image in Preview pane using top channel
- change channel to color - works as expected 
- change to LUT - works in Preview pane - button and slider change
- open image in Full Viewer - button on left has not changed to LUT
- open rendering settings panel - shows correct LUT button and slider
- change LUT - change accepted (re-open color picker and confirms) - button and slider in rendering settings panel not updated to new LUT selection  - see screenshot 1
- close color picker and save settings in Full Viewer - Preview settings on that channel revert to colour selected before LUT was selected - see screenshot 2

main 2

main 1

**Comment by jburel** [comment]
as indicated on gh-4767, backend on breaking so you will have to point to a different server in order to see the changes.

**Comment by dominikl** [comment]
Thanks @gusferguson . Could you give it another test today, with @jburel 's LUT backend PR in merge build? Expected behaviour for the channel color button now is, that there's no background, when a LUT is selected (to be match the web behaviour), only the slider shows the LUT colors.

**Comment by jburel** [comment]
I am really not convinced by the "list of luts" in any of the PRs.
We need to find a better solution
3 locations now that we have to update if we add an official lut (insight/web/figure)

**Comment by dominikl** [comment]
Ideal would be a server side method to request a Map of LUT names and icons with one call.

**Comment by gusferguson** [comment]
@dominikl - happy to retest.

**Comment by gusferguson** [comment]
@dominikl 

Tested using OMERO.insight-5.3.0-m2-1057-599ba47-ice36-b387-mac user-3 read-only-1

In Preview pane, Full Viewer selection of LUT behaves as expected for selection and display of LUTs - it has same issue with ?alpha channel rendering as described in:
https://github.com/ome/figure/pull/160#issuecomment-237483773

Copy and paste of LUTs not working in Preview pane or Full Viewer
- channel with LUT is toggled off when settings pasted on an image
- also if LUT is set on image paste normal colour for that channel does not over-write it
- getting some very strange behaviour when pasting - sometimes unchanged channels are also toggling off on paste after one channel has been set to LUT

Save button not always activating after LUT is changed in channel - seems to happen mainly when one LUT selection is saved and then another selected - intermittent

Sometimes selection of channel with LUT causes channel above to toggle off.

The intermittent issues seem to get worse with progressive changes.

I will keep my Insight open if you can’t replicate so you can see.

Consistent issue:
In Preview Pane, select LUT for channel
- button to left of slider changes to generic white
- open image in Full Viewer (before or after saving)
- button on left in full viewer remains colour selected before LUT applied
- in rendering settings panel of Full Viewer LUT button is white as expected
- selecting LUT in Full Viewer rendering settings does not change the left hand button colour
- see screenshot 1

main 1

**Comment by jburel** [comment]
@gusferguson: https://github.com/openmicroscopy/openmicroscopy/pull/4753/commits/d57c42e21cd1978c480c0ab64306613f7ec2480e fixes the "override" issue

**Comment by jburel** [comment]
mentioned to @dominikl: Lut selection missing from color picker in main viewer.
The problem occurs when opening the color picker from the main window i.e. icon on left side of image or when clicking on the channel itself

**Comment by jburel** [comment]
Another issue: when I select a LUT the heigh of the slider is not preserved. The slider is smaller than when I use a color

**Comment by dominikl** [comment]
Found some more places where the channel color is used, and where the lookup table has to be taken into account. Also hopefully fixed the synchronization issues between preview and full image viewer. Found another problem - can't figure out, how to fix it, yet: The split view image shows the channel color not the lookup table, but apparently only if not all channels are activated.

**Comment by gusferguson** [comment]
@dominikl 

Tested using OMERO.insight-5.3.0-m2-1084-bf9bc35-ice36-b390-win eel user-3 read-only-1

LUT not blocking out channels below it any more - behaves as expected.
LUT selection in Preview being shown when image opened in Full Viewer - behaves as expected

With efforts to synchronise between main window and full viewer is a change in behaviour from 5.2.4 that we need to be clear that we want:

- if change colour of channel to LUT or colour in Full Viewer this reflected in Preview pane immediately
- toggling channels on or off not reflected in Preview pane
- if I toggle a channel on/off in Full Viewer, then change a colour, then the channel toggles are reflected in Preview
- all of these hold for changes in Preview reflecting in Full Viewer
- in 5.2.4 both channel colour changes and channel toggle off or on are only reflected in Preview pane when Save is done.
- the same is true in 5.2.4 for changes in Preview pane reflecting in open Full Viewer - only when Save is done.

The channel colour change is therefore acting like a save.
My opinion is that the 5.2.4 behaviour is the correct one and we should follow that. 
@pwalczysko  @jburel  - what do you feel?

Copy and paste of rendering settings still not working with LUTs.

**Comment by jburel** [comment]
we should keep the behaviour of 5.2.x for channel color changes: lut is the same.

**Comment by pwalczysko** [comment]
@gusferguson : Just as @jburel says https://github.com/openmicroscopy/openmicroscopy/pull/4764#issuecomment-238502829. No changes, no difference made between LUTs and colors.

**Comment by pwalczysko** [comment]
Bearing in mind the desired situation, see https://github.com/openmicroscopy/openmicroscopy/pull/4764#issuecomment-238508462, there is couple of discrepancies:
  - the LUT selection (unlike the color selection) in either Full Viewer or Preview has undesired consequences in that the 
       - Copy, Paste of Rnd to another image does not work unless you Saved the change (as reported by @gusferguson above)
       - Undo and Redo of a change to LUT is not possible (this change does not seem to be in the cache at all to go back to)
       - if you Saved changes on an image with LUTs and do Copy and Paste on another image, behaviour is erratic (pastes old Colors only, instead of pasting the new LUT)

Bug: The Save icon button after LUT is applied in Full Viewer on Split view produces balck thumbnail in Rnd panel of Full Viewer (see screenshot below).
screen shot 2016-08-09 at 11 26 38

Bug: Split view in Full viewer does not react to changes of Rnd settings (moving the sliders etc.), the split channels in the main Full Viewer section ignore your sliding of the Rnd sliders.

**Comment by dominikl** [comment]
With respect to:
- if change colour of channel to LUT or colour in Full Viewer this reflected in Preview pane immediately
- toggling channels on or off not reflected in Preview pane

Can we go the other way, and also synchronize the toggle channels action between full viewer and preview? The settings are not saved unless the save button is clicked. I really don't know how I could otherwise deal with that, because parts of the UI are synchronized (because they are shared between preview and full viewer) and some are not e. g. the various channel buttons and where else the channel color icon appears (Acquisition tabs, etc.).

**Comment by gusferguson** [comment]
@dominikl 

In my opinion that would be OK if that is the easiest way to do it.

An issue I was having with that way that will have to be addressed:
- Full Viewer open - make change in Preview pane - save in Preview
- close Full Viewer - prompts for another save.

**Comment by pwalczysko** [comment]
Agree with @gusferguson https://github.com/openmicroscopy/openmicroscopy/pull/4764#issuecomment-238520164

**Comment by dominikl** [comment]
- Ok, turned out it was easier, to revert my previous changes, and leave the state like it was in 5.2.x. Preview and full viewer are now again *not* synchronized with respect to 'volatile' rendering settings changes, but they should show the same state, after the 'save' button has been clicked.
There's still one part left where the channel color icon gets not updated, this is the channel tab on the acquisition pane. It's initialized correctly, but after changing the color (or lut) it still shows the original color/lut icon. But this didn't work before either.

- Undo/redo and copy/paste should now work with respect to lookup tables.

**Comment by pwalczysko** [comment]
working with the custom built insight OMERO.insight-5.3.0-m3-ice36-SNAPSHOT-mac which includes all the commits here:

  - Copy Paste and Undo Redo works now with the LUTs (and colors) as expected

**Comment by pwalczysko** [comment]
working with the custom built insight OMERO.insight-5.3.0-m3-ice36-SNAPSHOT-mac which includes all the commits here:

Bug:  Preview gets frozen and unusable when a Full Viewer workflow has been performed.
   - open Preview on an image
   - open Full Viewer of this image as well
   - open the Rnd pane on the Full Viewer
   - make a change of LUT (or color)
   - hit Save icon in Full Viewer
   - close Full Viewer -> observe you get a question asking you whether or not to Save changes (should not be there actually)
   - answer "No" to the question
   - observe that the Full Viewer closed as expected, but
   - Preview is now not displayed at all (see screenshot below) and you cannot manipulate the controls in the Preview pane

screen shot 2016-08-10 at 10 34 14

**Comment by dominikl** [comment]
I've seen that bug yesterday on your laptop, but I still can't replicate it, followed exactly the steps you mentioned. Only got the question dialog about saving rendering settings, which yes, shouldn't pop up in that case.

**Comment by pwalczysko** [comment]
Bug: Thumbnail in Full Viewer goes black  every time.after Save icon is pressed
screen shot 2016-08-10 at 10 47 04

**Comment by jburel** [comment]
Anything in the log?

**Comment by dominikl** [comment]
Can't find out what causes this issue. It happens randomly, best chances to replicate is, view image in preview panel and full viewer at the same time. Modify one channel's visibility (toggle on or off, doesn't matter), don't save, but also change the lookup table of another channel, now save. Thumbnails get reloaded, but the thumbnail in the fullviewer is likely to be black.

Stacktrace:

```
 serverStackTrace = "ome.conditions.OptimisticLockException: You are not authorized to change the update event for ome.model.display.Thumbnail:Id_9 from ome.model.meta.Event:Id_1251 to ome.model.meta.Event:Id_1204
                        You may need to reload the object before continuing.
                        	at ome.security.basic.OmeroInterceptor.managedEvent(OmeroInterceptor.java:1072)
                        	at ome.security.basic.OmeroInterceptor.checkManagedDetails(OmeroInterceptor.java:843)
                        	at ome.security.basic.OmeroInterceptor.resetDetails(OmeroInterceptor.java:389)
                        	at ome.security.basic.OmeroInterceptor.onFlushDirty(OmeroInterceptor.java:184)
                        	at org.hibernate.event.def.DefaultFlushEntityEventListener.invokeInterceptor(DefaultFlushEntityEventListener.java:372)
                        	at org.hibernate.event.def.DefaultFlushEntityEventListener.handleInterception(DefaultFlushEntityEventListener.java:349)
                        	at org.hibernate.event.def.DefaultFlushEntityEventListener.scheduleUpdate(DefaultFlushEntityEventListener.java:287)
                        	at org.hibernate.event.def.DefaultFlushEntityEventListener.onFlushEntity(DefaultFlushEntityEventListener.java:155)
                        	at org.hibernate.event.def.AbstractFlushingEventListener.flushEntities(AbstractFlushingEventListener.java:219)
                        	at org.hibernate.event.def.AbstractFlushingEventListener.flushEverythingToExecutions(AbstractFlushingEventListener.java:99)
                        	at org.hibernate.event.def.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:50)
                        	at org.hibernate.impl.SessionImpl.flush(SessionImpl.java:1208)
                        	at ome.logic.UpdateImpl.afterUpdate(UpdateImpl.java:295)
                        	at ome.logic.UpdateImpl.doAction(UpdateImpl.java:311)
                        	at ome.logic.UpdateImpl.doAction(UpdateImpl.java:302)
                        	at ome.logic.UpdateImpl.saveObject(UpdateImpl.java:108)
                        	at ome.services.ThumbnailBean.retrieveThumbnailAndUpdateMetadata(ThumbnailBean.java:1096)
                        	at ome.services.ThumbnailBean.getThumbnail(ThumbnailBean.java:1072)
                        	at sun.reflect.GeneratedMethodAccessor1253.invoke(Unknown Source)
                        	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                        	at java.lang.reflect.Method.invoke(Method.java:498)
                        	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
                        	at ome.security.basic.EventHandler.invoke(EventHandler.java:154)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.tools.hibernate.SessionHandler.doStateful(SessionHandler.java:216)
                        	at ome.tools.hibernate.SessionHandler.invoke(SessionHandler.java:200)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:108)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:249)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:121)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
                        	at com.sun.proxy.$Proxy97.getThumbnail(Unknown Source)
                        	at sun.reflect.GeneratedMethodAccessor1253.invoke(Unknown Source)
                        	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                        	at java.lang.reflect.Method.invoke(Method.java:498)
                        	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
                        	at ome.security.basic.BasicSecurityWiring.invoke(BasicSecurityWiring.java:93)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.services.blitz.fire.AopContextInitializer.invoke(AopContextInitializer.java:43)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
                        	at com.sun.proxy.$Proxy97.getThumbnail(Unknown Source)
                        	at sun.reflect.GeneratedMethodAccessor1264.invoke(Unknown Source)
                        	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                        	at java.lang.reflect.Method.invoke(Method.java:498)
                        	at ome.services.blitz.util.IceMethodInvoker.invoke(IceMethodInvoker.java:172)
                        	at ome.services.throttling.Callback.run(Callback.java:56)
                        	at ome.services.throttling.InThreadThrottlingStrategy.callInvokerOnRawArgs(InThreadThrottlingStrategy.java:56)
                        	at ome.services.blitz.impl.AbstractAmdServant.callInvokerOnRawArgs(AbstractAmdServant.java:140)
                        	at ome.services.blitz.impl.ThumbnailStoreI.getThumbnail_async(ThumbnailStoreI.java:125)
                        	at sun.reflect.GeneratedMethodAccessor1263.invoke(Unknown Source)
                        	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                        	at java.lang.reflect.Method.invoke(Method.java:498)
                        	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
                        	at omero.cmd.CallContext.invoke(CallContext.java:78)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
                        	at com.sun.proxy.$Proxy98.getThumbnail_async(Unknown Source)
                        	at omero.api._ThumbnailStoreTie.getThumbnail_async(_ThumbnailStoreTie.java:114)
                        	at omero.api._ThumbnailStoreDisp.___getThumbnail(_ThumbnailStoreDisp.java:541)
                        	at omero.api._ThumbnailStoreDisp.__dispatch(_ThumbnailStoreDisp.java:875)
                        	at IceInternal.Incoming.invoke(Incoming.java:221)
                        	at Ice.ConnectionI.invokeAll(ConnectionI.java:2536)
                        	at Ice.ConnectionI.dispatch(ConnectionI.java:1145)
                        	at Ice.ConnectionI.message(ConnectionI.java:1056)
                        	at IceInternal.ThreadPool.run(ThreadPool.java:395)
                        	at IceInternal.ThreadPool.access$300(ThreadPool.java:12)
                        	at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:832)
                        	at java.lang.Thread.run(Thread.java:745)
                        "
    serverExceptionClass = "ome.conditions.OptimisticLockException"
    message = "You are not authorized to change the update event for ome.model.display.Thumbnail:Id_9 from ome.model.meta.Event:Id_1251 to ome.model.meta.Event:Id_1204
               You may need to reload the object before continuing."
    backOff = 0
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at IceInternal.BasicStream.createUserException(BasicStream.java:2773)
	at IceInternal.BasicStream.access$300(BasicStream.java:14)
	at IceInternal.BasicStream$EncapsDecoder10.throwException(BasicStream.java:3292)
	at IceInternal.BasicStream.throwException(BasicStream.java:2285)
	at IceInternal.OutgoingAsync.throwUserException(OutgoingAsync.java:399)
	at omero.api.ThumbnailStorePrxHelper.end_getThumbnail(ThumbnailStorePrxHelper.java:1642)
	at omero.api.ThumbnailStorePrxHelper.getThumbnail(ThumbnailStorePrxHelper.java:1513)
	at omero.api.ThumbnailStorePrxHelper.getThumbnail(ThumbnailStorePrxHelper.java:1500)
	at org.openmicroscopy.shoola.env.data.views.calls.ThumbnailLoader.loadThumbail(ThumbnailLoader.java:141)
	at org.openmicroscopy.shoola.env.data.views.calls.ThumbnailLoader.access$7(ThumbnailLoader.java:115)
	at org.openmicroscopy.shoola.env.data.views.calls.ThumbnailLoader$2.doCall(ThumbnailLoader.java:236)
	at org.openmicroscopy.shoola.env.data.views.BatchCall.doStep(BatchCall.java:144)
	at org.openmicroscopy.shoola.util.concur.tasks.CompositeTask.doStep(CompositeTask.java:226)
	at org.openmicroscopy.shoola.env.data.views.CompositeBatchCall.doStep(CompositeBatchCall.java:126)
	at org.openmicroscopy.shoola.util.concur.tasks.ExecCommand.exec(ExecCommand.java:165)
	at org.openmicroscopy.shoola.util.concur.tasks.ExecCommand.run(ExecCommand.java:276)
	at org.openmicroscopy.shoola.util.concur.tasks.AsyncProcessor$Runner.run(AsyncProcessor.java:91)
	at java.lang.Thread.run(Thread.java:745)
Exception in thread "Thread-267"
```

**Comment by jburel** [comment]
it looks like a problem with the thumbnail service and the luts
We will need a test to pin that down.

**Comment by dominikl** [comment]
Last commits:
- Updated the lookup table names and icons to include the new LUTs, see #4795
- Reduced the number of calls to the Thumbnail service. That should minimize the chances to hit an `OptimisticLockException` which causes the black thumbnail in the preview panel. I still couldn't find out the real cause for this. The trigger seems to be multiple simultaneous requests of the same thumbnail. No idea why a `get` request would perform an update (which might run into the lock issue).
- A bugfix suggested by @gusferguson : If the icon for a lookup table was missing, the layout of the lookup table list looked quite messy. Now an empty 'placeholder' icon is used, to preserve the layout.
- When a image was shown in the preview and in the full viewer at the same time, modifying and saving the rendering settings in the full viewer, still triggered the "Save Data" dialog, when closing the full viewer. That should be fixed now, too.

**Comment by gusferguson** [comment]
@dominikl 

Tested using OMERO.insight-5.3.0-m2-1162-597cc31-ice36-b406-mac user-3 read-only-1

All colour picker and LUT functionality behaves as expected.
Could not check alignment of labels of LUT without images as only short list of LUTs  - all with images was being displayed.
Save dialog no longer appearing.

Good to merge if can check LUTs with no image thumbnail are behaving themselves.

**Comment by dominikl** [comment]
Would it be possible to upload a new lookup table to eel just for testing how the client's handle an 'unknown' lookup table? Maybe via a temporary PR, might be the easiest option? /cc @jburel

**Comment by jburel** [comment]
You can use the upload as admin. (e.g. via cli or insight)
You can modify an existing text lut, change its name and upload it

**Comment by dominikl** [comment]
@gusferguson On eel there's now an additional lookup table called `glow 2`. That one doesn't have an icon in the clients.

**Comment by gusferguson** [comment]
@dominikl 

LUT without image working as expected.

Good to merge.

**Comment by pwalczysko** [comment]
All seems good here although https://github.com/openmicroscopy/openmicroscopy/pull/4764#issuecomment-238818378 definitely persists. Happy to merge, possibly https://github.com/openmicroscopy/openmicroscopy/pull/4764#issuecomment-238818378 can be filed as a long-standing mystery.

**Comment by jburel** [comment]
@pwalczysko:probably fixed by commit	 https://github.com/openmicroscopy/openmicroscopy/pull/4764/commits/b63d91c75722e8981093b32fa0362a01da13153b
but let's keep an eye on it

**Comment by pwalczysko** [comment]
@jburel :  No, https://github.com/openmicroscopy/openmicroscopy/pull/4764/commits/b63d91c75722e8981093b32fa0362a01da13153b does not fix it. I was able to repeat just today, the bug is present.

**Comment by dominikl** [comment]
With b63d91c it's less likely to hit the bug, but yes, you're right, it's still there. I get it quite often with eel, hardly get it when testing with local server. The trigger seems to be that the same thumbnail is requested twice simultaneously when the preview tab and the full viewer is open. Not sure how I could avoid that. And actually it shouldn't cause a problem in first place.

**Comment by jburel** [comment]
Can you add to https://trello.com/c/9iAe3Sg9/95-lut-support
known issues since will need to look at it in a follow up PR

**Comment by pwalczysko** [comment]
Added to the card.

**Comment by jburel** [comment]
@pwalczysko thanks

**Pull request merged**